xdsl/parser/attribute_parser.py

```diff
@@ -670,61 +672,174 @@ def _parse_optional_builtin_parametrized_attr(self) -> Attribute | None:
+            # perform check on just the flattened shapes for hex attribute
+            assert shape is not None
+            # For splat attributes any shape is fine
+            if shape == []:
+                return
+            if type_num_values != shape[0]:
+                self.raise_error(
+                    f"Shape mismatch in dense literal. Expected {type_num_values} "
+                    f"elements from the type, but got {shape[0]} elements."
+                )
+
+        def _check_empty_dense_attr(type_num_values: int):
+            if type_num_values != 0:
+                self.raise_error(
+                    "Expected at least one element in the "
+                    "dense literal, but got None"
+                )
+
+        self.parse_punctuation("<", " in dense attribute")
+        # The flattened list of elements
         values: list[AttrParser._TensorLiteralElement]
         # The dense shape.
-        # If it is `None`, then there is no values.
-        # If it is `[]`, then this is a splat attribute, meaning it has the same
-        # value everywhere.
+        #   If it is `None`, then there is no values.
+        #   If it is `[]`, then this is a splat attribute, meaning it has the
+        #   same value everywhere.
         shape: list[int] | None
+        # A hex_string:
+        #   If a hex_string is present it is a string.
+        #   Otherwise it is None
+        hex_string: str | None = None
+
         if self._current_token.text == ">":
             values, shape = [], None
         else:
-            values, shape = self._parse_tensor_literal()
+            hex_string = self.parse_optional_str_literal()
+            if hex_string is not None:
+                # Can not determine values without type yet
+                values, shape = [], None
+            else:
+                # Expect a tensor literal instead
+                values, shape = self._parse_tensor_literal()
         self.parse_punctuation(">", " in dense attribute")
 
-        # Parse the dense type.
-        self.parse_punctuation(":", " in dense attribute")
-        type = self.expect(self.parse_optional_type, "Dense attribute must be typed!")
-
-        # Check that the type is correct.
-        if not isa(
-            type,
-            RankedVectorOrTensorOf[IntegerType]
-            | RankedVectorOrTensorOf[IndexType]
-            | RankedVectorOrTensorOf[AnyFloat],
-        ):
-            self.raise_error(
-                "Expected vector or tensor type of " "integer, index, or float type"
-            )
-
-        # Check that the shape matches the data when given a shaped data.
+        type = _parse_dense_type()
         type_shape = list(type.get_shape())
-        num_values = math.prod(type_shape)
-
-        if shape is None and num_values != 0:
-            self.raise_error(
-                "Expected at least one element in the " "dense literal, but got None"
-            )
```

Comment:
upon deeper inspection, this check makes sense (if you encounter a `dense<>` there can not be any shape in your type (when does this happen??) ), but the message really doesn't.
Should I change it?